feat: middleware and virtual routes (#7254)
withastro/docs: src/content/docs/en/guides/middleware.mdx

---
import PackageManagerTabs from '~/components/tabs/PackageManagerTabs.astro';

**Middleware** allows you to intercept requests and responses and inject behaviors dynamically every time a page or endpoint is about to be rendered. This rendering occurs at build time for all prerendered pages, but occurs when the route is requested for pages rendered on demand.

> Middleware also allows you to set and share request-specific information across endpoints and pages by mutating a `locals` object that is available in all Astro components and API endpoints.

Reviewer: ```suggestion
Middleware also allows you to set and share request-specific information across endpoints and pages by mutating a `locals` object that is available in all Astro components and API endpoints. This object is available even when this middleware runs at build time, giving your static routes access to information during the lifecycle of a request.
```
What do you think about incorporating this idea here?

> Middleware also allows you to set and share request-specific information across endpoints and pages by mutating a `locals` object that is available in all Astro components and API endpoints. This object is available even when this middleware runs at build time, giving your static routes access to stored information during the lifecycle of a request.

Author: ```suggestion
Middleware also allows you to set and share request-specific information across endpoints and pages by mutating a `locals` object that is available in all Astro components and API endpoints. This object is available even when this middleware runs at build time.
```
Not blocking, but this could be confusing - there is no "real" request at build time to derive information from.

Reviewer: Love it, thank you!

Middleware is available in both SSG and SSR Astro projects.

### Timing of middleware execution

> By default, astro builds your project into a static site. In this [mode](https://docs.astro.build/en/basics/rendering-modes/), the middleware runs during the build as well. This allows the pages access to [`locals` object](#storing-data-in-contextlocals). Additionally, the static files can be served immediately since no server-side code is run on-request, ensuring fast response times.

Author: Users ask "why not run middleware only on-request for static pages". So I tried to give the 2 reasons here:
- you'll need `Astro.locals`, and
- you probably don't want server-side code to run on-request if it can run ahead of time for performance.

Reviewer: OK, so I see this less about timing and more about "why would I run middleware on static pages?"

In that framing, I can see adding something here. Let me try something!

> By default, astro builds your project into a static site. In this [mode](https://docs.astro.build/en/basics/rendering-modes/), the middleware runs during the build as well. This allows the pages access to [`locals` object](#storing-data-in-contextlocals). Additionally, the static files can be served immediately since no server-side code is run on-request, ensuring fast response times.

Author: ```suggestion
By default, astro builds your project into a static site. In this [mode](/en/basics/rendering-modes/), the middleware runs during the build as well. This allows the pages access to [`locals` object](#storing-data-in-contextlocals). Additionally, the static files can be served immediately since no server-side code is run on-request, ensuring fast response times.
```

Middleware is available in both SSG and SSR Astro projects.

:::note
On all prerendered pages, the middleware runs only during the **build process** to create static pages and does not run in the deployed website. This applies to all pages in a `static` (SSG) project. This also includes prerendered pages in `hybrid` mode (default) and any pages in `server` mode with `export const prerender = true`, including 404 and 500 pages.

> For on-demand rendered pages in `server` (default) or `hybrid` (with `export const prerender = false`) mode, the middleware runs in the deployed website when the route is requested. If a route is not found, the middleware will run for the 404 page, unless it is prerendered.

Author: ```suggestion
For on-demand rendered pages in `server` (default) or `hybrid` (with `export const prerender = false`) mode, the middleware runs in the deployed website when the route is requested. If a route is not found, the middleware will run for the 404 page, unless it is prerendered. Some adapters may show a platform-specific 404 page instead, in which case the middleware will not run.
```

auth response
validation response
```

> ### Handling 404 pages

Author: ```suggestion
## Error pages
```
It is not as recognizable as having "404" in the heading, but we don't have enough to say about 500 pages to give it its own section.

`###` to `##` because it isn't a subsection of "Chaining middleware".

Reviewer: Agreed!

> Middleware will run for an on-demand rendered route while Astro attempts to render a 404 page. However, it is up to the [adapter](/en/guides/server-side-rendering/) to decide whether that code runs. Some adapters may serve a platform-specific error page instead.

Author: ```suggestion
If your custom 404 page is on-demand rendered, or if astro's default 404 page is being used, middleware will run for requests that did not match a route. However, it is up to the [adapter](/en/guides/server-side-rendering/) to decide whether that code runs. Some adapters may serve a platform-specific error page instead.
```

Author: I feel a little hesitant using the term "default 404 page" because it is just a blank white page. There may not be any avoiding it, and that's fine too.

Reviewer: I feel like my sentence is a cleaner user-facing statement -- middleware will run when...

What about something closer to:

> Middleware will attempt to run for all on-demand rendered pages, even when a matching route cannot be found. This includes Astro's default (blank) 404 page and any custom 404 pages. However, ...

Reviewer: Updated your suggestion with min to commit. See what you think of this

> Middleware will run for an on-demand rendered route while Astro attempts to render a 404 page. However, it is up to the [adapter](/en/guides/server-side-rendering/) to decide whether that code runs. Some adapters may serve a platform-specific error page instead.

Author: ```suggestion
Middleware will run for an on-demand rendered route while Astro attempts to render a 404 page. However, it is up to the [adapter](/en/guides/server-side-rendering/) to decide whether that code runs. Some adapters may serve a platform-specific error page instead.

If a page fails to render due to an error, astro will attempt to render your custom 500 page. If the error page is set to render on-demand, the middleware will run a second time for that request.
```

Reviewer: I'm trying to figure out how *important* it is to add these two paragraphs. 500 is generic error, right? Does it *matter* that middleware runs if the page 500s? What practical value is this to the author, and would they change anything about how they write their code knowing this? In which situations would it make a difference?

Similarly, if a 500 page is being served,how much does it *matter* that they don't have `Astro.locals`? 

I'm going by the "not everything is true belongs in docs" principle, and want to include things that will affect people's decisions, and that they need to know to build properly. I'm going to need your help to figure out if these fit that criteria.

Author: Not that important.

The most important point would be that 500 page could miss `locals` _sometimes_. However, even that I don't feel too strongly about - it came up only once.

> Middleware will attempt to run for all on-demand rendered pages, even when a matching route cannot be found. This includes Astro's default (blank) 404 page and any custom 404 pages. However, it is up to the [adapter](/en/guides/server-side-rendering/) to decide whether that code runs. Some adapters may serve a platform-specific error page instead.

Reviewer: ```suggestion
Middleware will attempt to run for all on-demand rendered pages, even when a matching route cannot be found. This includes Astro's default (blank) 404 page and any custom 404 pages. However, it is up to the [adapter](/en/guides/server-side-rendering/) to decide whether that code runs. Some adapters may serve a platform-specific error page instead.

Middleware will also attempt to run before serving a 500 error page, including a custom 500 page, unless the server error occured in the execution of the middleware itself. If your middleware does not run successfully, then you will not have access to `Astro.locals` to render your 500 page.
```
Does something like this capture it?
